refactor: Duplicate and update primitives made available.

Kelketek opened the pull request:
## Description

This pull request makes a couple of changes to the xblock handler in the CMS. These changes add a handful of utility functions and modify the existing ones to make reuse of existing blocks easier. With these changes, it is possible to copy an entire section from one course to another, and then later refresh that section, and all of its children, without destroying the blocks next to it.

The existing `_duplicate_block` function was modified to have a `shallow` keyword to avoid copying children, and the `update_from_source` function was added to make it easy to copy attributes over from one block to another. These functions can be used alongside `copy_from_template` in the modulestore to copy over blocks and their children without requiring them to be within any particular container (other than a library or course root)-- thus allowing library-like inclusion without the library content block. This is especially useful for cases like copying sections rather than unit content.

The actual code for copying sections into new courses is not included, as the UX for such a feature is not yet decided, but prototypical Django admin functionality is being built in a separate Django app repository.

From the perspective of the user, this PR should change nothing. From the perspective of the developer, it should make things much easier.

## Testing instructions

The unit tests should prove this is working correctly and doesn't impact existing functionality. However a companion Django app is being developed which leverages these features and can be made available to test as well.

## Deadline

None

## Misc

Internal/private-ref: BB-7295

Kelketek committed:
refactor: Duplicate and update primitives made available.

openedx-webhooks commented:
Thanks for the pull request, @Kelketek!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

Agrendalath commented on cms/djangoapps/contentstore/views/block.py:
```diff
@@ -878,47 +878,87 @@ def _move_item(source_usage_key, target_parent_usage_key, user, target_index=Non
+    return duplicate_metadata, asides_to_create
+
+
+def update_from_source(*, source_block, destination_block, user_id):
```
Why do we want to disable positional arguments here?

Kelketek replied:
Because mixing them up-- especially the first two-- seems easy and consequential.

Agrendalath commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -3473,3 +3497,20 @@ def test_self_paced_item_visibility_state(self, store_type):
+        source_chapter = BlockFactory(parent=course, category='chapter', display_name='Source Chapter')
+        destination_chapter = BlockFactory(parent=course, category='chapter', display_name='Destination Chapter')
+        update_from_source(source_block=source_chapter, destination_block=destination_chapter, user_id=user.id)
+        self.assertEqual(self.store.get_item(destination_chapter.location).display_name, 'Source Chapter')
```
Would it make sense to verify that the `metadata`, `definition_data` and `asides` are updated correctly too?

Kelketek replied:
This turned out to be harder to do than I expected, but I've added tests for this now as well as a variant that verifies that they get properly clobbered as they should be if content exists there already.

Agrendalath commented on xmodule/modulestore/store_utilities.py:
```diff
@@ -104,3 +106,23 @@ def get_draft_subtree_roots(draft_nodes):
+    """
+    Return a new reproducible block ID for a given root, source block, and destination parent.
+
+    This is useful for ensuring that 'refreshable' blocks always have the same ID when updated.
```
Would you mind adding a more verbose description of this use case to this docstring? This is the only occurrence of the "refreshable" term in the whole `edx-platform`.

Kelketek replied:
Done!

Kelketek commented:
@Agrendalath This is ready for another look.

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: checked that duplicating the blocks is still working; checked that [this plugin](https://github.com/open-craft/section-to-course/pull/1) can use new functions
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Kelketek commented:
@bradenmacdonald This is ready for you.

Kelketek commented:
@bradenmacdonald Forgot Piotr was a CC and can move this forward-- but heads up on the changes in any case.

bradenmacdonald commented on cms/djangoapps/contentstore/views/block.py:
```diff
@@ -878,47 +878,88 @@ def _move_item(source_usage_key, target_parent_usage_key, user, target_index=Non
+    dest_usage_key=None,
+    display_name=None,
+    shallow=False,
+    is_child=False
```
Can you please document all these arguments? It doesn't say what `shallow` or `is_child` do, nor the purpose of `dest_usage_key`.

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -67,7 +67,7 @@
     _xblock_type_and_display_name,
     add_container_page_publishing_info,
-    create_xblock_info,
+    create_xblock_info, duplicate_block, update_from_source,
```
This should be giving a lint error? Should be one import per line here.

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -786,6 +786,29 @@ def verify_name(source_usage_key, parent_usage_key, expected_name, display_name=
 
+    def test_shallow_duplicate(self):
+        """
+        Test that shallow_duplicate creates a new block.
```
```suggestion
        Test that duplicate_block(..., shallow=True) can duplicate a block but ignores its children.
```

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -786,6 +786,29 @@ def verify_name(source_usage_key, parent_usage_key, expected_name, display_name=
+        self.assertEqual(len(source_chapter.get_children()), 1)
+        destination_course = CourseFactory()
+        destination_location = duplicate_block(
+            parent_usage_key=destination_course.location, duplicate_source_usage_key=source_chapter.location,
```
```suggestion
            parent_usage_key=destination_course.location,
            duplicate_source_usage_key=source_chapter.location,
```
nit - should be consistently one argument per line.

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -3473,3 +3496,111 @@ def test_self_paced_item_visibility_state(self, store_type):
+        super().setUp()
+        key_store = DictKeyValueStore()
+        field_data = KvsFieldData(key_store)
+        self.runtime = TestRuntime(services={'field-data': field_data})
```
Do you need to set up a `TestRuntime` here, with its own keystore and everything? I feel like it's better to just use the normal modulestore, which `ModuleStoreTestCase` sets up for you - then it's essentially the same in the test environment as it is in prod.

Kelketek replied:
@bradenmacdonald I'm not sure how to use that with the tools I have available for testing asides. This is based on how asides were tested elsewhere in this module, where I borrowed it from. The AsideTest class is based on XBlockAside, which requires a generated def_id, usage_id a runtime argument, apparently from which the def_id and usage_id are generated. I'm not fully certain what all of this is, just that it's required.

If there's some better way to do it I'm happy to, but if that's the case it would be better to change it elsewhere in these tests, too.

bradenmacdonald replied:
Asides are a half-baked feature and I think only used by one group at MIT for one very specific thing? I don't think it's necessary to test them here. We can just say that this particular functionality doesn't support asides.

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -3473,3 +3496,111 @@ def test_self_paced_item_visibility_state(self, store_type):
+        usage_id = self.runtime.id_generator.create_usage(def_id)
+
+        aside = AsideTest(scope_ids=ScopeIds('user', 'html', def_id, usage_id), runtime=self.runtime)
+        aside.field11 = 'html_new_value1'
```
Do we need to test asides? I don't think they were ever fully implemented as a feature.

Kelketek replied:
Unless we plan to remove them, new code should continue support for them, I think. Otherwise we have a degrading internal feature that we have to repair before we can use, if we ever decide to finish it. That is to say, maybe asides should get a DEPR ticket, but I'd rather my code work with them so long as they're here.

bradenmacdonald replied:
I have asked on Slack #dev about the status.

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -3473,3 +3496,111 @@ def test_self_paced_item_visibility_state(self, store_type):
+
+    def create_source_block(self, course):
+        """
+        Create a chapter with all the fixings.
```
You could also use `ToyCourseFactory` which has some content "with all the fixings" ready to go?

Kelketek replied:
That's out of pattern with the rest of the tests in this module, and I don't think it'd be as clear for these tests in particular, since I'll be testing specific attributes that would otherwise be implicitly created.

bradenmacdonald replied:
OK, that's totally fine. Just didn't want you writing code if there was an existing thing you could use instead.

bradenmacdonald commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -3473,3 +3496,111 @@ def test_self_paced_item_visibility_state(self, store_type):
+    @XBlockAside.register_temp_plugin(AsideTest, 'test_aside')
+    def test_update_clobbers(self):
+        """
+        Verify that our update clobbers everything.
```
Could you please be a little more explicit about what this is supposed to do? :p

bradenmacdonald commented on xmodule/modulestore/store_utilities.py:
```diff
@@ -104,3 +106,26 @@ def get_draft_subtree_roots(draft_nodes):
             yield draft_node
+
+
+def derived_key(courselike_source_key, block_key, parent):
```
`parent` is fairly ambiguous - parent of the source or of the destination? UsageKey? XBlock? BlockKey? I'd prefer sticking with the name `new_parent_block_key` or even better `dest_parent: BlockKey` with a type hint.

bradenmacdonald commented on xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2438,23 +2437,14 @@ def _copy_from_template(
-            )
-            new_block_id = hashlib.sha1(unique_data.encode('utf-8')).hexdigest()[:20]
-            new_block_key = BlockKey(block_key.type, new_block_id)
+            new_block_key = derived_key(src_course_key, block_key, new_parent_block_key)
```
Moving `derived_key` to a separate function does seem a bit cleaner, but what's the motivation here? It's only used in this one place still as far as I can see.

bradenmacdonald commented on cms/djangoapps/contentstore/views/block.py:
```diff
@@ -878,47 +878,88 @@ def _move_item(source_usage_key, target_parent_usage_key, user, target_index=Non
+    return duplicate_metadata, asides_to_create
+
+
+def update_from_source(*, source_block, destination_block, user_id):
```
If this is a python API that you intend to use from other python code and which doesn't itself provide a REST API, it doesn't belong in `views/`. It should be in `api.py` or similar. (And maybe it should be in the modulestore code somewhere instead of contentstore? Not really sure on that.) I know the existing code in this file doesn't follow such conventions very clearly, but we can at least not make it worse. OEP-49 for reference.

bradenmacdonald reviewed (commented):
From what I can see, a lot of the diff is just moving some things to separate functions.

So is this what's actually changed?
* Added `shallow` to `duplicate_block`
* Turned `duplicate_block` from internal method to public API (should perhaps be moved out of views then?)
* Added a new `update_from_source` python API, but it's not used by the platform


> These functions can be used alongside copy_from_template

So there are three functions, `duplicate_block`, `update_from_source`, and `copy_from_template`. Can you give an example of how you're planning to use them? I feel like it may be a little unclear the differences between them and when/how to decide which to use, so I'd like to see that made clearer within the code/docstrings. I'm also wondering if we need three separate high level APIs when it seems like it's all aiming to do one thing: copy blocks into a destination and update them if they already exist there.

> without requiring them to be within any particular container (other than a library or course root)

I was a little confused at first because I thought you meant "without requiring the **source block** to be in any particular container like a library". But I see you mean "without requiring them to be placed into a library content block or any other "special" parent container".

So my question is: how does the currently available API require the use of a particular container for the destination, and how does this PR remove that requirement?

Kelketek commented:
@bradenmacdonald I've addressed a few of your notes and responded to some comments. I've added more to the docstrings to make it clearer when they're to be used.

The remaining issue is a much larger one and I'm not sure how to approach it. That issue being 'moving these functions into their own API module', which sounds simple, but there is already an `api` module in contentstore, so either I would need to rename that module, or put this new api module within the existing api package, or call it something else. The simplest solution would be to stuff them in some other module that, while it wouldn't follow OEP-45, would at least not be in the middle of the views. This might be the best option.

Another alternative would be to place it in the modulestore package. There is not currently an api package in the modulestore, and this seems like an odd place to start-- especially since we didn't make one as part of BD-13. However, it is an option. Another potential option might be to turn these functions into function that live on the modulestore itself, but, oddly enough, there's already a `copy` function within the modulestore. So we'll have copy, copy_from_template, and duplicate_block. Not ideal.

I don't know that any approach is going to be perfect, but I wanted to see if you had any other ideas. Otherwise I'll probably try just moving the functions out of the views to another module, even if it's not named `api.py`.

bradenmacdonald commented:
@Kelketek You're right, this is an older part of the code that's definitely not following the conventions I'm referring to, so it's difficult to integrate with it in that way. I'm actually fine with what you're doing here, though I think that putting it into any file outside of `views` is better (maybe `cms/djangoapps/contentstore/utils.py` ?). I don't want to put a larger refactor onto your shoulders just because you're adding one little python method.

What's important is that it's clear who will be using this method. As it is, if it looks like a helper method in `views/block.py` but none of the views in that file are using it, I'd be inclined to think it's a "forgotten" helper method that's no longer used, and delete it. So that's something a good docstring can help with. As it stands I would see this method as "ready for deletion".

Kelketek commented:
@bradenmacdonald OK-- I think this is ready now. Assuming we don't want to remove the tests for asides, that is-- and it sounds like maybe we don't, since now more than one group is using them and there appears to be no interest in deprecating them.

bradenmacdonald commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -765,3 +775,201 @@ def update_course_discussions_settings(course_key):
+    under the supplied dest_usage_key.
+
+    If shallow is True, does not copy children. This function calls itself recursively,
+    and will set the is_child flag to True when dealing with recursed child blocks.
```
Should be: "If shallow is True, does not copy children. **Otherwise,** this function calls itself recursively..."

bradenmacdonald commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -765,3 +775,201 @@ def update_course_discussions_settings(course_key):
+):
+    """
+    Duplicate an existing xblock as a child of the supplied parent_usage_key, for placement
+    under the supplied dest_usage_key.
```
"for placement under the supplied dest_usage_key." is confusing. Saying "under" sounds like `dest_usage_key` is to be the parent of the new copy.

I would suggest: "Duplicate an existing xblock as a child of the supplied parent_usage_key. You can optionally specify what usage key the new duplicate block will use via `dest_usage_key`."

bradenmacdonald commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -765,3 +775,201 @@ def update_course_discussions_settings(course_key):
+    This function is meant to be imported by pluggable django apps looking to manage duplicated
+    sections of a course. It is placed here for lack of a more appropriate location, since this
+    code has not yet been brought up to the standards in OEP-45.
+    """
```
Nice docstring, thanks!

Kelketek commented:
@bradenmacdonald Thanks-- I've rebased this and updated the docstrings you mentioned.

Kelketek commented:
@bradenmacdonald @Agrendalath Now that this is approved, is there anything left to do on my end or can one of you merge this in?

openedx-webhooks commented:
@Kelketek 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/32058&entry.1484655318=https://github.com/Kelketek&entry.752974735=2023-04-11+23:10&entry.1917517419=2023-05-25+13:58&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
